### ISSUE TITLE
Support for purging version history of a resource.

### ISSUE THREAD

**Issue by topherbullock** [concourse/concourse#145]
Currently there is no way to clean out the version history of a resource outside of re-creating it under a different name. This would be useful if the state of a resource changes outside of concourse and the reference history is no longer true, or for rolling back the version resource to the initial state.

**Comment by jamesjoshuahill** [comment]
+1

**Comment by pivotal-topher-bullock** [comment]
@vito # make it happen

**Comment by devdavidkarlsson** [comment]
#913 duplicate-ish

**Comment by pivotal-topher-bullock** [comment]
@vito
# make 
# it
# happen

**Comment by devdavidkarlsson** [comment]
This is such a horrible bug for me 🤕  just disabling the version leaves the jobs stuck in pending...

**Comment by craigfurman** [comment]
@vito is there any chance of this being prioritized soon? Especially after https://github.com/concourse/concourse/issues/372 was closed ;)

**Comment by jamesjoshuahill** [comment]
Please please please 🐬

**Comment by joshzarrabi** [comment]
I also I agree that this would be a good thing.

**Comment by vito** [comment]
Questions I have around this (not shutting it down, we just need to define the semantics):

* What happens to the data builds have referencing particular versions as inputs / outputs?
  * Do we want these truly gone in that case, or somehow marked "deleted"? How does this relate to the later story around resources supporting a `delete` action (#534)? If at all?
* Is it really just 'delete all' or would 'delete individual/range' be useful? (Are there implications for which route we take?)
  * We'll need to figure out the UI/UX for this based on the decision we make here.
* Is the need for this really pointing at a gap in the story around resources? Should it be *possible* to get in this state? For example, changing a resource's config could clear things out, or `check` could be used to "sync" with the source of truth, or maybe changing a resource's config *forces* that sync, etc. etc.

The risk in adding this is similar to our "knobs" anti-pattern (which we'll hopefully have a doc for soon), i.e. having controls that require constant/periodic attention from users in order for Concourse to continue functioning correctly.

**Comment by devdavidkarlsson** [comment]
Marked as deleted and not used in new builds would work in my case, If it then would force the pull of a new resource that is.

I thought the power button on the resource versions would have this effect but they did not.

**Comment by jamesjoshuahill** [comment]
Hi @vito, thank you for these excellent questions!

I think delete and delete range (e.g. all) actions would have been useful on the occasions I have wished for purge. Here are two scenarios that might help discussion:

**Testing jobs that bump semver in a version resource**

Once you've bumped the version you cannot go back :o( In this case, only the latest version is needed. So the actions I imagined were:

- pause the pipeline
- set initial version in version resource
- purge version resource history
- restart the pipeline

If I deleted the semvers used during testing then I presume I would not be able to use them again. So I would have to remember to set an arbitrarily high test version before I started testing jobs to avoid collisions with deleted versions later. Maybe it would be simpler to purge, i.e. forget, those versions of the resource and the builds associated with them?

**Tag moved on a git resource**

Once a worker has pulled a git resource it cannot detect a tag when it is moved. Whilst moving git tags raises other problems, it does happen. In this case, I wanted to flush the cache of the git resource, which is more like the `sync` idea you suggested.

**Comment by spikymonkey** [comment]
We just been stung by this after cloning a pipeline to start building from a new release series branch (`1.5.x`), and accidentally leaving a version resource pulling from a version file in the old git branch (master). As the versions in master advanced to `2.0.0` and beyond, they were pulled into our new `1.5.x` builds... and as others have found, there's no good way back from this. We've ended up renaming the version resource and `input_mapping` it in all of our tasks, which adds unnecessary complexity to our pipeline. It'd be great to have a way to fix this kind of issue that don't involve making permanent changes to the pipeline.

**Comment by krishicks** [comment]
I had a lot of trouble with this while developing a resource; the versions weren't coming back in the right order which meant each time I wanted to try something different I had to create a new resource.

I also had to create a new tag each time because there's no way I can tell when Concourse has actually picked up the new version. Docker itself seems to be slow at reporting new versions after a push, so there's some bit of time I'd have to wait before I could even do the rename of the resource to get a fresh set of versions.

**Comment by arbourd** [comment]
I came across this today and found a somewhat temporary solution -- destroy the entire pipeline ;) Obviously, this won't work for everyone.

**Comment by clarafu** [comment]
This issue is resolved through the completion of #2386 because resource versions are no longer tied to a resource name, but rather through it's config. For example, each time the source of a resource is changed, it will produce new versions. If we had a pipeline with a resource:
```
resources:
- name: some-resource
  type: git
  source:
    repository: concourse/concourse
```
and we set the pipeline again after changing the source: 
```
resources:
- name: some-resource
  type: git
  source:
    repository: concourse/other-concourse
```
the version history would be reset, containing only the versions for `concourse/other-concourse`. 

Essentially, we did not add support for purging resource history, but instead have resource versions depend on it's resource config (resource type and source) instead of the resource name.

**Comment by kdvolder** [comment]
Might be useful to someone else stumbling in here looking for information.... As this wasn't immediately obvious to me. 

I didn't really want to change the resource configuration, I just wanted to reset version history that I had poluted by doing a bunch of testing. 

Fortunately, in such a case, it is actually possible now to reset history by making a trivial/meaningless change to the config. For example I added a 'cache_buster' attribute to the source like so:

```
resources:
- name: some-resource
  type: git
  source:
    repository: concourse/concourse
    cache_buster: whatever
```

THe cache_buster attribute has no meaning to the resource implementation so its not a 'real' change, but it does trigger a version reset.
